File: providers/Elasticsearch/src/CorpusScanner.php
<?php

namespace HawaiianSearch;

class CorpusScanner {
    private EmbeddingClient $embeddingClient;
    private MetadataCache $metadataCache;
    private array $hawaiianWords; // Now a hash set
    private bool $dryrun;

    // Patterns for text analysis
    private const DIACRITIC_PATTERN = '/[āĀēĒīĪōŌūŪ\']/u';
    private const WORD_PATTERN = '/\b\w+\b/u';

    public function __construct(ElasticsearchClient $esClient, array $options = []) {
        $this->embeddingClient = $esClient->getEmbeddingClient();
        $this->metadataCache = new MetadataCache($esClient, $options);
        $this->hawaiianWords = isset($options['hawaiianWords']) ? $options['hawaiianWords'] : [];
        $this->dryrun = isset($options['dryrun']) ? $options['dryrun'] : false;
    }

    /**
     * Hash a sentence text to create consistent IDs
     */
    public static function hashSentence(string $text): string {
        return md5(strtolower(trim($text)));
    }

    /**
     * Normalize Hawaiian word for dictionary lookup
     */
    public static function normalizeWord(string $word): string {
        // Remove okina and apostrophes
        $word = str_replace(["'", "'"], "", $word);
        
        // Convert macrons to plain vowels
        $macrons = ['ā', 'ē', 'ī', 'ō', 'ū', 'Ā', 'Ē', 'Ī', 'Ō', 'Ū'];
        $plain = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'];
        $word = str_replace($macrons, $plain, $word);
        
        return strtolower(trim($word));
    }

    /**
     * Calculate Hawaiian word ratio for text
     */
    public function calculateHawaiianWordRatio(string $text): float {
        if (empty(trim($text))) {
            return 0.0;
        }
        
        preg_match_all(self::WORD_PATTERN, $text, $matches);
        $words = $matches[0];
        $wordCount = count($words);
        
        if ($wordCount === 0) {
            return 0.0;
        }

        $hawaiianWordCount = 0;

        foreach ($words as $word) {
            // Check for diacritical marks first (strong indicator of Hawaiian)
            if (preg_match(self::DIACRITIC_PATTERN, $word)) {
                $hawaiianWordCount++;
            } else {
                // Check against Hawaiian word dictionary
                $normalizedWord = self::normalizeWord($word);
                if (isset($this->hawaiianWords[$normalizedWord])) {
                    $hawaiianWordCount++;
                }
            }
        }
        
        return $hawaiianWordCount / $wordCount;
    }

    /**
     * Calculate basic entity count using simple pattern matching
     */
    public function calculateEntityCount(string $text): int {
        $entityCount = 0;
        
        // Count proper nouns (capitalized words that aren't at sentence start)
        $sentences = preg_split('/[.!?]+/', $text);
        foreach ($sentences as $sentence) {
            $sentence = trim($sentence);
            if (empty($sentence)) continue;
            
            preg_match_all('/\b[A-Z][a-z]+\b/', $sentence, $matches);
            $capitalizedWords = $matches[0];
            
            // Skip the first word if it's capitalized (likely sentence start)
            $words = preg_split('/\s+/', trim($sentence));
            if (!empty($words) && !empty($capitalizedWords)) {
                $firstWord = $words[0];
                $capitalizedWords = array_filter($capitalizedWords, function($word) use ($firstWord) {
                    return $word !== $firstWord;
                });
            }
            
            $entityCount += count($capitalizedWords);
        }
        
        // Count dates (simple patterns)
        $entityCount += preg_match_all('/\b\d{1,4}[\/\-]\d{1,2}[\/\-]\d{2,4}\b/', $text);
        
        // Count numbers that might be significant
        $entityCount += preg_match_all('/\b\d{4}\b/', $text); // Years
        
        return $entityCount;
    }

    /**
     * Compute boilerplate score (quality metric)
     * Lower score = better quality (0.0 to 1.0)
     */
    public function computeBoilerplateScore(string $text, int $entityCount): float {
        $score = 0.0;
        
        // Penalize very short text
        if (strlen($text) < 40) {
            $score += 0.5;
        }
        
        // Penalize text with no entities
        if ($entityCount === 0) {
            $score += 0.5;
        }
        
        // Additional heuristics for boilerplate detection
        $text_lower = strtolower($text);
        
        // Penalize repetitive patterns
        if (preg_match('/(.{20,})\1{2,}/', $text)) { // Repeated 20+ char patterns
            $score += 0.3;
        }
        
        // Penalize common boilerplate phrases
        $boilerplatePatterns = [
            '/copyright|all rights reserved/i',
            '/click here|more info/i', 
            '/terms of service|privacy policy/i',
            '/subscribe|newsletter/i'
        ];
        
        foreach ($boilerplatePatterns as $pattern) {
            if (preg_match($pattern, $text)) {
                $score += 0.2;
                break; // Don't double-penalize
            }
        }
        
        return round(min($score, 1.0), 2);
    }

    /**
     * Full sentence analysis with all metrics
     */
    public function analyzeSentence(string $text, string $docId = '', array $existingMetadata = null): array {
        $hash = self::hashSentence($text);

        // Start with existing metadata or default values
        $metadata = $existingMetadata ?: ['frequency' => 0];
        
        // Increment frequency count
        $metadata['frequency'] = ($metadata['frequency'] ?? 0) + 1;

        // Calculate basic metrics locally
        preg_match_all(self::WORD_PATTERN, $text, $matches);
        $words = $matches[0];
        $wordCount = count($words);
        
        $entityCount = $this->calculateEntityCount($text);
        $hawaiianWordRatio = $this->calculateHawaiianWordRatio($text);
        $boilerplateScore = $this->computeBoilerplateScore($text, $entityCount);

        // Update metadata with all calculated metrics
        $metadata['sentence_hash'] = $hash;
        $metadata['length'] = strlen($text);
        $metadata['entity_count'] = $entityCount;
        $metadata['word_count'] = $wordCount;
        $metadata['hawaiian_word_ratio'] = $hawaiianWordRatio;
        $metadata['boilerplate_score'] = $boilerplateScore;

        // Track document IDs and positions where this sentence appears
        if (!isset($metadata['metadata'])) {
            $metadata['metadata'] = ['doc_ids' => [], 'positions' => []];
        }
        
        // Add current document and position if provided and not already present
        if (!empty($docId) && !in_array($docId, $metadata['metadata']['doc_ids'])) {
            $metadata['metadata']['doc_ids'][] = $docId;
            $metadata['metadata']['positions'][] = 0; // Position would need to be passed in
        }

        // Save to the metadata cache if not in dry run mode
        if (!$this->dryrun) {
            $this->metadataCache->set($hash, $metadata);
        }

        return $metadata;
    }

    /**
     * Legacy method for backward compatibility - calls the enhanced analyzeSentence
     */

    public function finish(): void {
        $this->metadataCache->finish();
    }


    public function setHawaiianWords(array $words): void {
        $this->hawaiianWords = $words;
    }
}

File: providers/Elasticsearch/python/corpus_scanner.py
import argparse
import hashlib
from collections import defaultdict
from spacy import load as load_spacy
from elasticsearchclient import ElasticsearchDB
from metadata_cache import MetadataCache  # ⬅️ New cache class

class CorpusScanner:
    def __init__(self, config, recreate=False, verbose=False, save_interval=10, client=None, dryrun=False):
        self.config = config
        self.es = client if client else ElasticsearchDB(config)
        self.dryrun = dryrun
        self.nlp = load_spacy("en_core_web_sm")
        self.metric_keys = ["length", "frequency", "boilerplate_score", "entity_count", "word_count", "hawaiian_word_ratio"]
        self.sentence_counts = defaultdict(int)
        self.verbose = verbose
        self.save_interval = save_interval
        self.hawaiian_words = self.load_hawaiian_words()
        print(f"Read {len(self.hawaiian_words)} words into Hawaiian dictionary")

        # 🧠 Read-write cache backed by metadata index
        self.metadata_cache = MetadataCache(self.es, max_size=10000, flush_interval=60)

    @staticmethod
    def _normalize_word(word):
        # Remove okina and apostrophe, and convert macrons to plain vowels
        return word.lower().replace("‘", "").replace("'", "").replace("ā", "a").replace("ē", "e").replace("ī", "i").replace("ō", "o").replace("ū", "u")

    def load_hawaiian_words(self):
        try:
            with open("hawaiian_words.txt", "r") as f:
                all_headwords = set()
                for line in f:
                    # 1. Split by comma for multiple headwords on one line
                    comma_parts = [p.strip() for p in line.split(',')]
                    for part in comma_parts:
                        # 2. Split by whitespace for multi-word headwords
                        words = part.split()
                        for word in words:
                            # 3. Skip 'a' and 'i'
                            if word.lower() not in ['a', 'i']:
                                all_headwords.add(word)
                
                # 4. Normalize all collected words and return as a set for uniqueness
                return {self._normalize_word(word) for word in all_headwords}
        except FileNotFoundError:
            print("Warning: hawaiian_words.txt not found. Hawaiian word ratio will not be calculated.")
            return set()

    @staticmethod
    def hash_sentence(text):
        return hashlib.md5(text.strip().lower().encode("utf-8")).hexdigest()

    def compute_boilerplate_score(self, text, entity_count):
        score = 0
        if len(text) < 40:
            score += 0.5
        if entity_count == 0:
            score += 0.5
        return round(min(score, 1.0), 2)

    def analyze_sentence(self, text, existing_metadata=None):
        h = CorpusScanner.hash_sentence(text)
        meta = existing_metadata or {}

        meta["frequency"] = meta.get("frequency", 0) + 1
        doc = self.nlp(text)
        entity_count = len(doc.ents)
        
        words = text.split()
        word_count = len(words)
        hawaiian_word_count = 0
        diacritic_chars = "āĀēĒīĪōŌūŪ‘"

        for word in words:
            if any(char in word for char in diacritic_chars):
                hawaiian_word_count += 1
            else:
                normalized_word = self._normalize_word(word)
                if normalized_word in self.hawaiian_words:
                    hawaiian_word_count += 1
        
        hawaiian_word_ratio = hawaiian_word_count / word_count if word_count > 0 else 0

        meta.update({
            "length": len(text),
            "entity_count": entity_count,
            "boilerplate_score": self.compute_boilerplate_score(text, entity_count),
            "word_count": word_count,
            "hawaiian_word_ratio": hawaiian_word_ratio
        })

        self.metadata_cache.set(h, meta)
        return meta

    def process_sentence(self, text, doc_id):
        h = CorpusScanner.hash_sentence(text)
        existing_meta = self.metadata_cache.get(h)
        meta = self.analyze_sentence(text, existing_meta)

        if self.verbose:
            print(f"[METADATA] {h} → {text[:100]}...")

    def scan(self):
        total_processed = 0
        for doc in self.es.get_all_sentences():
            doc_id = doc["doc_id"]
            source = doc["_source"].get("sourcename", "UNKNOWN")
            sentences = doc["_source"].get("sentences", [])
            added = 0

            for s in sentences:
                text = s.get("text")
                self.process_sentence(text, doc_id)
                added += 1

            print(f"[SCAN] Doc: {doc_id}, Source: {source}, Sentences processed: {added}")
            total_processed += 1

        self.finish()

    def finish(self):
        self.metadata_cache.finish()

    def fetch_metadata(self):
        metadata = self.es.fetch_all_metadata()
        for h, m in metadata.items():
            print(f"{h}: {m}")

    def show_sentence_metadata(self, doc_id):
        doc = self.es.get_document(doc_id)
        if not doc or "_source" not in doc or "sentences" not in doc["_source"]:
            print(f"⚠️ Document {doc_id} not found or missing sentence data.")
            return

        sentences = doc["_source"]["sentences"]
        print(f"📘 Showing stored metadata for {len(sentences)} sentences in doc {doc_id}:\n")

        for i, s in enumerate(sentences):
            text = s.get("text", "").strip()
            if not text:
                continue

            h = CorpusScanner.hash_sentence(text)
            meta = self.metadata_cache.get(h)

            print(f"📄 Sentence [{i}]")
            print(f"Text: \"{text}\"")
            print(f"Hash: {h}")

            if meta:
                for key in self.metric_keys:
                    print(f"{key.capitalize()}: {meta.get(key, '—')}")
            else:
                print("⚠️ No stored metadata found for this sentence.")
            print("-" * 40)

    def rehash_sentences(self):
        sentence_docs = self.es.get_all_sentences()
        for doc in sentence_docs:
            doc_id = doc["doc_id"]
            sentences = doc["_source"].get("sentences", [])
            updated = []

            for i, s in enumerate(sentences):
                if "sentence_hash" in s:
                    continue

                text = s["text"]
                h = CorpusScanner.hash_sentence(text)

                if "vector" not in s:
                    full_doc = self.es.get_document(doc_id)
                    full_sent = full_doc["_source"]["sentences"][i]
                    full_sent["sentence_hash"] = h
                    updated.append(full_sent)
                else:
                    s["sentence_hash"] = h
                    updated.append(s)

            if updated:
                self.es.update_document(doc_id, {"doc": {"sentences": updated}})
                print(f"🔄 Rehashed doc: {doc_id}")

# ── Entry Point ─────────────────────────────────────
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--recreate", action="store_true")
    parser.add_argument("--fetch", action="store_true")
    parser.add_argument("--rehash", action="store_true")
    parser.add_argument("--verbose", action="store_true")
    parser.add_argument("--save_interval", type=int, default=10)
    parser.add_argument("--show_sentences", type=int, help="Display full sentence metadata for a specific doc_id")
    args = parser.parse_args()

    config = { "COLLECTION_NAME": "hawaiian_hybrid" }
    scanner = CorpusScanner(config, recreate=args.recreate, verbose=args.verbose, save_interval=args.save_interval)

    if args.recreate:
        scanner.es.recreate_metadata_index()
    elif args.rehash:
        scanner.rehash_sentences()
    elif args.fetch:
        scanner.fetch_metadata()
    elif args.show_sentences:
        scanner.show_sentence_metadata(args.show_sentences)
    else:
        scanner.scan()
